### DISCUSSION TITLE
How can I import dll files?

### DISCUSSION THREAD

**Discussion by BetaFoprhoton** [cc-tweaked/CC-Tweaked#1963] [Q&A] [answered]
Recently, I want to import some dll files for data processing and other purposes, but I found that the package.loadlib function originally used by Lua for importing dll files does not seem to exist in Craft OS. But the original package.preload and package.path still exist. I don't know if this is for safety reasons or my own problem. So is there a way to import external dll files in Craft OS?
![issue](https://github.com/user-attachments/assets/6a0e9a57-9952-4aba-a8c6-9709a0f48dac)

**Comment by SquidDev** [comment, marked as answer]
CC uses a Java-based Lua implementation, and so doesn't support native libraries. As you mention, it also wouldn't be especially safe (especially on servers), as native code wouldn't be subject to CC's sandbox. 

Depending on what the library did, it should be possible to replicate its functionality in pure Lua.

**Reply by BetaFoprhoton** [reply]
Okay, so libs can be run on a server then indirectly transfer data to the CC computer through the internet? But this will be affected by the web‘s stability.

**Reply by SquidDev** [reply]
Yep, if CC's not fast enough to do the processing, you can definitely offload it to an external server.

**Comment by zyxkad** [comment]
Native library are depending on the OS, which means dll on windows, so on linux, and dylib on macos. And CraftOS should not depending on the platform it runs on.

Safety is also a problem, if you can load dynamic library, you can execute any command you want.

The `package.path` is for load Lua package when you invoking `require`, not for low level library.

**Comment by TimeBather** [comment]
If this DLL file is just used by your owned private Minecraft server, a simple way is create a mod that import it using JNI and exposure to the lua VM.
